File: lib/autorouter-pipelines/AutoroutingPipeline9_PreloadedTraceGraph/filter-pipeline9-drc-errors-against-baseline.ts
type DrcError = Record<string, unknown>

const DRC_ERROR_ID_KEYS = [
  "pcb_trace_error_id",
  "pcb_error_id",
  "pcb_via_trace_clearance_error_id",
  "pcb_pad_trace_clearance_error_id",
] as const

const isMissingConnectionError = (error: DrcError): boolean =>
  typeof error.pcb_trace_error_id === "string" &&
  error.pcb_trace_error_id.startsWith("missing_connection_")

const normalizePreparedTraceIds = (
  value: string,
  originalTraceIdByPreparedTraceId: ReadonlyMap<string, string>,
): string => {
  let normalized = value
  const aliases = [...originalTraceIdByPreparedTraceId].sort(
    ([left], [right]) => right.length - left.length,
  )
  for (const [preparedTraceId, originalTraceId] of aliases) {
    normalized = normalized.replaceAll(preparedTraceId, originalTraceId)
  }
  return normalized
}

const getViaClearanceErrorIdentity = (
  error: DrcError,
  originalTraceIdByPreparedTraceId: ReadonlyMap<string, string>,
): string | undefined => {
  const traceIds = [
    ...(typeof error.pcb_trace_id === "string" ? [error.pcb_trace_id] : []),
    ...(Array.isArray(error.pcb_trace_ids)
      ? error.pcb_trace_ids.filter(
          (traceId): traceId is string => typeof traceId === "string",
        )
      : []),
  ]
    .map((traceId) =>
      normalizePreparedTraceIds(traceId, originalTraceIdByPreparedTraceId),
    )
    .filter(
      (traceId, traceIndex, allTraceIds) =>
        allTraceIds.indexOf(traceId) === traceIndex,
    )
    .sort()
  const centerCandidate =
    error.center && typeof error.center === "object"
      ? (error.center as Record<string, unknown>)
      : error.pcb_center && typeof error.pcb_center === "object"
        ? (error.pcb_center as Record<string, unknown>)
        : undefined
  const center =
    typeof centerCandidate?.x === "number" &&
    typeof centerCandidate.y === "number"
      ? { x: centerCandidate.x, y: centerCandidate.y }
      : undefined
  const netRelation =
    typeof error.pcb_via_pair_net_relation === "string"
      ? error.pcb_via_pair_net_relation
      : undefined

  // Sequential via IDs depend on trace iteration order, so they cannot prove
  // that a candidate violation was inherited. Missing stable metadata is kept
  // repairable instead of risking a false baseline match.
  if (traceIds.length === 0 || !center || !netRelation) return undefined
  return `pcb_via_clearance_error:${JSON.stringify({ traceIds, center, netRelation })}`
}

const getDrcErrorIdentity = (
  error: DrcError,
  originalTraceIdByPreparedTraceId: ReadonlyMap<string, string>,
): string | undefined => {
  const errorType = String(error.type ?? error.error_type ?? "unknown")
  if (errorType === "pcb_via_clearance_error") {
    return getViaClearanceErrorIdentity(error, originalTraceIdByPreparedTraceId)
  }
  for (const idKey of DRC_ERROR_ID_KEYS) {
    const errorId = error[idKey]
    if (typeof errorId === "string") {
      return `${errorType}:${normalizePreparedTraceIds(errorId, originalTraceIdByPreparedTraceId)}`
    }
  }

  const identityFields = Object.fromEntries(
    Object.entries(error)
      .filter(
        ([key, value]) =>
          (key.endsWith("_id") || key.endsWith("_ids")) &&
          (typeof value === "string" || Array.isArray(value)),
      )
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([key, value]) => [
        key,
        typeof value === "string"
          ? normalizePreparedTraceIds(value, originalTraceIdByPreparedTraceId)
          : value,
      ]),
  )
  return `${errorType}:${JSON.stringify(identityFields)}`
}

/** Removes DRC violations already present in the supplied prerouted board. */
export const filterPipeline9DrcErrorsAgainstBaseline = <
  TError extends DrcError,
>({
  errors,
  baselineErrors,
  originalTraceIdByPreparedTraceId = new Map(),
}: {
  errors: TError[]
  baselineErrors: DrcError[]
  originalTraceIdByPreparedTraceId?: ReadonlyMap<string, string>
}): TError[] => {
  const baselineErrorIdentities = new Set(
    baselineErrors
      // A missing connection describes unfinished routing, not an inherited
      // geometric violation. Pipeline9 must continue to repair it even when
      // the same finding exists before the candidate routes are added.
      .filter((error) => !isMissingConnectionError(error))
      .map((error) => getDrcErrorIdentity(error, new Map())),
  )
  baselineErrorIdentities.delete(undefined)
  return errors.filter((error) => {
    const identity = getDrcErrorIdentity(
      error,
      originalTraceIdByPreparedTraceId,
    )
    return identity === undefined || !baselineErrorIdentities.has(identity)
  })
}

File: rust/capacity-autorouter/src/autorouter_pipelines/autorouting_pipeline9_preloaded_trace_graph/filter_pipeline9_drc_errors_against_baseline.rs
use indexmap::{IndexMap, IndexSet};
use serde_json::{Value, json};

use json_bindings::js_json::js_json;

fn normalize_prepared_trace_ids(value: &str, aliases: &IndexMap<String, String>) -> String {
    let mut sorted: Vec<_> = aliases.iter().collect();
    sorted.sort_by_key(|entry| std::cmp::Reverse(entry.0.len()));
    let mut result = value.to_owned();
    for (prepared, original) in sorted {
        result = result.replace(prepared, original);
    }
    result
}

fn get_drc_error_identity(error: &Value, aliases: &IndexMap<String, String>) -> Option<String> {
    let error_type = error
        .get("type")
        .filter(|v| !v.is_null())
        .or_else(|| error.get("error_type").filter(|v| !v.is_null()))
        .and_then(Value::as_str)
        .unwrap_or("unknown");
    if error_type == "pcb_via_clearance_error" {
        let mut ids = IndexSet::new();
        if let Some(id) = error["pcb_trace_id"].as_str() {
            ids.insert(normalize_prepared_trace_ids(id, aliases));
        }
        if let Some(values) = error["pcb_trace_ids"].as_array() {
            for id in values.iter().filter_map(Value::as_str) {
                ids.insert(normalize_prepared_trace_ids(id, aliases));
            }
        }
        let mut ids: Vec<_> = ids.into_iter().collect();
        ids.sort();
        let center = if error["center"].is_object() {
            &error["center"]
        } else {
            &error["pcb_center"]
        };
        let (Some(x), Some(y), Some(relation)) = (
            center["x"].as_f64(),
            center["y"].as_f64(),
            error["pcb_via_pair_net_relation"].as_str(),
        ) else {
            return None;
        };
        if ids.is_empty() {
            return None;
        }
        return Some(format!(
            "pcb_via_clearance_error:{}",
            js_json(&json!({"traceIds": ids, "center": {"x":x,"y":y}, "netRelation":relation}))
        ));
    }
    for key in [
        "pcb_trace_error_id",
        "pcb_error_id",
        "pcb_via_trace_clearance_error_id",
        "pcb_pad_trace_clearance_error_id",
    ] {
        if let Some(id) = error[key].as_str() {
            return Some(format!(
                "{error_type}:{}",
                normalize_prepared_trace_ids(id, aliases)
            ));
        }
    }
    let mut fields: Vec<_> = error
        .as_object()
        .expect("DRC error object")
        .iter()
        .filter(|(key, value)| {
            (key.ends_with("_id") || key.ends_with("_ids"))
                && (value.is_string() || value.is_array())
        })
        .collect();
    fields.sort_by(|a, b| a.0.cmp(b.0));
    let mut identity = serde_json::Map::new();
    for (key, value) in fields {
        identity.insert(
            key.clone(),
            value
                .as_str()
                .map(|s| Value::from(normalize_prepared_trace_ids(s, aliases)))
                .unwrap_or_else(|| value.clone()),
        );
    }
    Some(format!(
        "{error_type}:{}",
        js_json(&Value::Object(identity))
    ))
}

pub fn filter_pipeline9_drc_errors_against_baseline(
    errors: &[Value],
    baseline: &[Value],
    aliases: &IndexMap<String, String>,
) -> Vec<Value> {
    let baseline_ids: IndexSet<_> = baseline
        .iter()
        .filter(|error| {
            !error["pcb_trace_error_id"]
                .as_str()
                .is_some_and(|id| id.starts_with("missing_connection_"))
        })
        .filter_map(|error| get_drc_error_identity(error, &IndexMap::new()))
        .collect();
    errors
        .iter()
        .filter(|error| {
            get_drc_error_identity(error, aliases).is_none_or(|id| !baseline_ids.contains(&id))
        })
        .cloned()
        .collect()
}
